feat(rslint_parser): Ban template literal after optional chain

Boshen opened the pull request:
## Summary
Template string is not allowed to be passed to tail position of optional chain.


## Test Plan
### Test262 comparison coverage results on ubuntu-latest
| Test result | `main` count | This PR count | Difference |
| :---------: | :----------: | :-----------: | :--------: |
| Total | 44951 | 44951 | 0 |
| Passed | 43841 | 43845 | ✅ ⏫ **+4** |
| Failed | 1109 | 1105 | ✅ ⏬ **-4** |
| Panics | 1 | 1 | 0 |
| Coverage | 97.53% | 97.54% | **+0.01%** |

<details><summary><b>:tada: Fixed (4):</b></summary>

```
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-null-optchain-template-string-esi.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-null-optchain-template-string.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-optchain-template-string-esi.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-optchain-template-string.js
```
</details>

Boshen commented on crates/rslint_parser/src/syntax/expr.rs:
```diff
@@ -724,15 +724,17 @@ fn subscripts(
 ) -> CompletedMarker {
     // test_err subscripts_err
-    // foo()?.baz[].
+    // foo()?.baz[].;
```
```
// test_err subscripts_err
// foo()?.baz[].
// BAR`b
```
This wrapped, and triggered the wrong error :-(

ematipico commented on crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast:
```diff
@@ -0,0 +1,204 @@
+  ┌─ template_after_optional_chain.js:3:1
+  │
+3 │ obj.func?.(args)`template`
+  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Are we able to apply the range of the single template literal?

Boshen replied:
This whole expression is called "Tagged Template", so using the whole range or the template range are both good. But the error message mentions "optional chain" so the whole range is better?

We can apply to the single template literal, but then we have to pass an extra argument to `parse_template_literal` 🤔

MichaReiser replied:
Nit: An option would be to capture the `lhs` range before calling into `parse_template_literal` and then add a secondary range to the diagnostic, pointing to the optional chain expression. But I'm not sure if it adds a lot of clarity or makes things even more confusing.

MichaReiser commented on crates/rslint_parser/src/syntax/expr.rs:
```diff
@@ -797,7 +803,19 @@ fn subscripts(
+                // obj.func?.(args)`template`
+                if is_optional_chain {
+                    p.error(
+                        p.err_builder("Invalid tagged template on optional chain")
```
Can we use something closer to TS?

> Tagged template expressions are not permitted in an optional chain.

Where it becomes clear of whats

MichaReiser replied:
Can we change the type of the whole expression to an `UNKNOWN_EXPRESSION`?

Boshen commented:
I have updated the error message and changed it to unknown expression.

> Nit: An option would be to capture the lhs range before calling into parse_template_literal and then add a secondary range to the diagnostic, pointing to the optional chain expression. But I'm not sure if it adds a lot of clarity or makes things even more confusing.

Looks like it'll get more confusing :-(
